### ISSUE TITLE
Exact type props and spreading passthrough props

### ISSUE THREAD

**Issue by Ricardo-Marques** [facebook/flow#5551]
When typing the props on a component that accepts some passthrough props to give to its child, I'd expect this to work

```javascript
import * as React from 'react';

type FirstComponentProps = {|
	a: number,
	b: number       
|}

class FirstComponent extends React.Component<FirstComponentProps> {}

type SecondComponentOwnProps = {|
	c: number
|}

type SecondComponentProps = {|
	...FirstComponentProps,
	...SecondComponentOwnProps
|}

class MySecondComponent extends React.Component<SecondComponentProps> {
	render () {
      	        const { c, ...props } = this.props 
    	        return <FirstComponent {...props} />
	}
}

const Node = <MySecondComponent a={1} b={2} c={3} />
```

But it errors with `props of React element 'FirstComponent'. Inexact type is incompatible with exact type`

https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQdMAnmFnAGLBSowDCJkAHZZBMAArEw6ALxwA3gB86ASGQAuOIICuIAEZYoAGhW6N2vQbhXrcOgoC+DXABs06Ttz4CIw0XCwAHjAiACbo2HgwAHT84D4iMAA8HjyxQgkSEFIAfPKOjCxsAMpYuD4hafGiAPIA7oKZUnCyiiq4Zjr6UHb5zKxwJWWCFd6+4pIy8krKUbMpXnFjjajGM7OD5ZVjdQ0TPU6uqOgAskwbw1sJ-kGh4VTRl6KJ5yOLGRO5ciqUw5YAFABKeR0GxWZRDHjyOC4QxwWZRMATOD2ZpwGAAC2AqARSJB1mUlBgWiggjgyS4qVGVzk8MRWVQKIA9NkVI58hD4AA5CAhNiyRKnF6PeDIaRyACMKN0YoATCjcGKAMxMllAA

Passing each prop individually works
```javascript
class MySecondComponent extends React.Component<SecondComponentProps> {
	render () {
    	        return <FirstComponent a={this.props.a} b={this.props.b} />
	}
}
```

But this does not seem like it would be the best solution.

Also attempted this to no avail.
```javascript
class MySecondComponent extends React.Component<SecondComponentProps> {
	render () {
      	        const { c } = this.props
      	        const passThrough = getRest(this.props, { c })
      	        return <FirstComponent {...passThrough} />
	}
}

function getRest<Props: {}, ToOmit: {}> (
  props: Props, toOmit: ToOmit
): $Rest<Props, ToOmit> {
	const rest = { ...props }
	Object.keys(toOmit).forEach(prop => delete rest[prop])
	return rest
}
```

https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQdMAnmFnAGLBSowDCJkAHZZBMAArEw6ALxwA3gB86ASGQAuOIICuIAEZYoAGhW6N2vQbhXrcOgoC+DXABs06Ttz4CIw0XCwAHjAiACbo2HgwAHT84D4iMAA8HjyxQgkSEFIAfPKOjCxsAMpYuD4hafGiAPIA7oKZUnCyiiq4Zjr6UHb5zKxwJWWCFd6+4pIy8krKUbMpXnFjjajGM7OD5ZVjdQ0TPU6uqOgAskwbw1sJ-kGh4VTRl6KJ5yOLGRO5ciqUw5YAFABKeR0GxWZRDHjyOC4OD2ZpwGAAC2AqCiYD2oOsEPgYDcABVEcQtABzRHw4lYGDYHh-JEotETQxQmH2AEg6zKSgwLRQQRwZJcVKjK5yWZo-GEiAkxFwgD02RUjny2LgADkICE2LJEqcXo94MhpHIAIxw3RGgBMcNwRoAzHKFXQCFpBPhgD44BSqVgeIllho5PYmXiINUQMAYAH7Lk-uz0VlUBplkyYKHw5G4CGwxG6ACNAASalJZOZtMRz5tHyQyiQlpwMXxpqOZTVXQAK1K0QA1lgmKhaWWYACogRoABRPCIv6N5q5TXOSlsGswADajYAumzOZSeXzl3R7EA

**Comment by idiostruct** [comment]
The spread operator creates an _unsealed_ object type which is incompatible with an exact type. Pass props explicitly (as you've noted) or use inexact types.

**Comment by Ricardo-Marques** [comment]
I see. How does that affect spreading props? [I see that this still errors.](https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQdMAnmFnAGLBSowDCJkAHZZBMAArEw6ALxwA3gB86ASGQAuOIICuIAEZYoAGhW6N2vQbhXrcOgoC+DXABs06Ttz4CIw0XCwAHjAiACbo2HgwAHT84D4iMAA8HjyxQgkSEFIAfPKOjCxsAMpYuD4hafGiAPIA7oKZUnCyiiq4Zjr6UHb5zKxwJWWCFd6+4pIy8krKUbMpXnFjjajGM7OD5ZVjdQ0TPU6uqOgAskwbw1sJ-kGh4VTRl6KJ5yOLGRO5ciqUw5YAFABKeR0GzKSgwLRQQRwZJcVKjK5yWZRADmWBg80e4yyqD+MAAFsBUFEwBMAfY4AB6bIqRz5AhaQT4YA+OBojFwhbpUTLOB-Uk4gEaTEInkTYFWcGQ6FfGzqOACqRRZDGGymBUTKK6EFwOlOHw8OAAOQgITYskSpxeWJQ0jkAEYKbo7QAmCm4O0AZgp1LoQA)

```javascript
class MySecondComponent extends React.Component<SecondComponentProps> {
	render () {
    	        return <FirstComponent {...getFirstComponentProps(this.props)} />
	}
}

function getFirstComponentProps (props): FirstComponentProps {
      return {
            a: props.a,
            b: props.b
      }
}
```

Really seems like this should work. It would allow the definition for the function that extracts the props the component cares about (`getFirstComponentProps`) to live next to the corresponding component (`FirstComponent`), rather than having to rely on passing each prop individually or using non-exact types.

**Comment by idiostruct** [comment]
Flow seems to have issues inferring that the result of a spread operation is exact (#2405). A working, temporary, unideal solution is to use the shape of FirstComponentProps:

```diff
- class FirstComponent extends React.Component<FirstComponentProps> {}
+ class FirstComponent extends React.Component<$Shape<FirstComponentProps>> {}
```

**Comment by Ricardo-Marques** [comment]
[Good enough for me](https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQdMAnmFnAGLBSowDCJkAHZZBMAArEw6ALxwA3gB86ASGQAuOIICuIAEZYoAGhW6N2vQbhXrcOgoC+DXABs06Ttz4CIw0XCwAHjAiACbo2HgwAHT84D4iMAA8ACQAygAWyKyJHjyxQgkSEFIAfCXyjowsbKlYuD4h+fGiAPIA7oJFUnCyiiq4Zjr6UHaVzKxwtfWCjd6+4pIy8krKUWu5XnHzXajGq2tTDU3z7Z2Lo06uqOgAskyHM8cJ-kGh4VTRT6IpGVlYiQ9ZltCosyvIVJQZpYABQASnBNmUlBgWiggjgOS4eTmzzkayiAHMsDANl8FsVUNCYOlgKgomBFrD7HAAPQlFSOSoELSCfDAHxwIkkrGbAqiHZwaEMimwjSknHixYIuDI1HouR0GzqODSqRRZDGGymHWLKK6TVwTlOHw8OAAOQgITYskSd0BZJQ0jkAEZmbovQAmZm4L0AZmZbLoQA), thanks!

**Comment by Ricardo-Marques** [comment]
Unfortunately this solution does not work when [no props are given](https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQdMAnmFnAGLBSowDCJkAHZZBMAArEw6ALxwA3gB86ASGQAuOIICuIAEZYoAGhW6N2vQbhXrcOgoC+DXABs06Ttz4CIw0XCwAHjAiACbo2HgwAHT84D4iMAA8ACQAygAWyKyJHjyxQgkSEFIAfCXyjowsbKlYuD4h+fGiAPIA7oJFUnCyiiq4Zjr6UHaVzKxwtfWCjd6+4pIy8krKUWu5XnHzXajGq2tTDU3z7Z2Lo06uqOgAskyHM8cJ-kGh4VTRT6IpGVlYiQ9ZltCosyvIVJQZpYABQASnBNmUlBgWiggjgOS4eTmzwA9CUVI5KgQtIJ8MAfHAAOZYGAbL4LYroaFgRawjT0nGiHYIuDI1HouR0GzqOCsplRZDGGymMWLKK6YVwIlOHw8OAAOQgITYskSd0BDJQ0jkAEZ7HBdCaAEwW3AmgDMFvxdCAA).

**Comment by idiostruct** [comment]
Are exact types a requirement? If spread props are handled explicitly within components, inexact types could be sufficient:


```jsx
type FooProps = {
  a: number,
  b: number
};

class A extends React.Component<FooProps> {
  render() {
    return <div data-a={this.props.a} data-b={this.props.b} />;
  }
};

type BarProps = {
  c: number
};

type BazProps = FooProps & BarProps;

class B extends React.Component<BazProps> {
  render() {
    return <A {...this.props} />
  }
}
    
<B a={1} b={2} c={3} />
<B a={1} b={2} c={3} d={4} />
```

**Comment by Ricardo-Marques** [comment]
Yes, exact types are worth explicitly passing each prop to the component in my opinion..
I am currently building a UI component library to be used by my entire team and getting errors in real time for mistyped prop names is invaluable.

**Comment by ggarek** [comment]
>The spread operator creates an unsealed object type which is incompatible with an exact type. Pass props explicitly (as you've noted) or use inexact types.

Meaning, that [this](https://flow.org/try/#0C4TwDgpgBAglC8UDeAfKBDAXFAdgVwFsAjCAJwBoojsBnYUgSxwHMoUBfAbgChRIoAQgmQZs+YmShdu3AGZ4cAY2AMA9jiiyAFAA9sMAJT7k3KFFIRgeUhqRQAdI51Se7IA) will never work in flow _(and there is no intension to support it/it is not possible for a reason)_?

Code
```js
type A = {| a: number, b: string |};

function f(x: A): A {
  return { ...x };
}
```

p.s.: the issue is about using exact object types with spread operator, so i made a minimum demo:

/cc @idiostruct

**Comment by idiostruct** [comment]
Your minimum demo will work if the object returned by `f` is frozen:

```typescript
type A = {| a: number, b: string |};

function f1(x: A): A {
  // `o` can be mutated making it incompatible with an exact type.
  const o = { ...x };
  o.c = 3;
  return o;
}

function f2(x: A): A {
  // `o` cannot be mutated making it compatible with an exact type.
  const o = Object.freeze({ ...x });
  return o;
}
```

**Comment by kevinbarabash** [comment]
> The spread operator creates an unsealed object type which is incompatible with an exact type. Pass props explicitly (as you've noted) or use inexact types.

What's the rationale for this?  `<div {...props}/>;` compiles to `React.createElement("div", props);` so it's not like a copy of props is actually being created.  Also, I would expect that if `obj` is sealed that `{...obj}` would also be sealed.

**Comment by rattrayalex-stripe** [comment]
Given the release of Flow 70 with support for object-rest-spread maintaining a seal, I would expect the same to work with jsx-rest-spread. However, it [appears not to](https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFAwCeYWcACsWOgLxwDeAPgA8AXOQCMZPgF8auCADtU8AGIQIcbgAoeAySLYQOASnUA+OAB4AJsABuJ7ZPMB6a3ZnzFcAELIo6uBpgeuyoxpxm5ipqPAB0cWCScE4mQA) (even with `Object.freeze` [backflips](https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFAwCeYWcACsWOgLxwDeAPgA8AXOQCMZPgF8auCADtU8AGIQIcbgAoeAySLYQOASnUA+OAB4AJsABuJ7ZPMB6a3ZnzFcAELIo6uBpgeuyoxpxmPLIK8GAATP4A8gBGAFZY+AB0RFhYAF5YWhlFYJKGANyUMACuUHIWKmo8RRlxknBOJpJAA)).

😞

**Comment by kevinbarabash** [comment]
@rattrayalex-stripe I'm glad to hear that `object-rest-spread` maintains sealed objects in the latest version.  Hopefully it won't too long before `jsx-rest-spread` does the same.

**Comment by Ricardo-Marques** [comment]
Seems like this is resolved in the latest 0.71 release!! Closing this issue, thanks everyone!
